src/api.ts

```diff
@@ -0,0 +1,216 @@
+    data?: T,
+    errors?: Array<GraphQLError>,
+    extensions?: { [key: string]: unknown }
+}
+
+declare class API {
+    /** HTTP GET METHOD */
+    static get<ResponseBody extends unknown>(params: APIParam, init?: FetchOptions): Promise<APIResponse<ResponseBody>>;
+    /** HTTP HEAD METHOD */
+    static head<ResponseBody extends unknown>(params: APIParam, init?: FetchOptions): Promise<APIResponse<ResponseBody>>;
+    /** HTTP DELETE METHOD */
+    static delete<ResponseBody extends unknown>(params: APIParam, init?: FetchOptions): Promise<APIResponse<ResponseBody>>;
+    /** HTTP OPTIONS METHOD */
+    static options<ResponseBody extends unknown>(params: APIParam, init?: FetchOptions): Promise<APIResponse<ResponseBody>>;
+    /** HTTP PUT METHOD */
+    static put<RequestBody extends JSONValue = JSONValue, ResponseBody extends unknown = unknown>(params: APIParam, init: UpdateOptionsJSON<RequestBody>): Promise<APIResponse<ResponseBody>>
+    static put<RequestBody extends string = string, ResponseBody extends unknown = unknown>(params: APIParam, init: UpdateOptionsString<RequestBody>): Promise<APIResponse<ResponseBody>>
+    /** HTTP POST METHOD */
+    static post<RequestBody extends JSONValue = JSONValue, ResponseBody extends unknown = unknown>(params: APIParam, init: UpdateOptionsJSON<RequestBody>): Promise<APIResponse<ResponseBody>>
+    static post<RequestBody extends string = string, ResponseBody extends unknown = unknown>(params: APIParam, init: UpdateOptionsString<RequestBody>): Promise<APIResponse<ResponseBody>>
+    /** HTTP PATCH METHOD */
+    static patch<RequestBody extends JSONValue = JSONValue, ResponseBody extends unknown = unknown>(params: APIParam, init: UpdateOptionsJSON<RequestBody>): Promise<APIResponse<ResponseBody>>
+    static patch<RequestBody extends string = string, ResponseBody extends unknown = unknown>(params: APIParam, init: UpdateOptionsString<RequestBody>): Promise<APIResponse<ResponseBody>>
+    /** Cancel HTTP Request */
+    static cancel: (request: Promise<unknown>, message: string) => void
+    /** */
+    static query: <GraphqlData extends JSONObject = JSONObject>({ document, variables }: { document: string, variables?: JSONObject }) => Promise<GraphqlResult<GraphqlData>>
+    static mutation: <GraphqlData extends JSONObject = JSONObject>({ document, variables }: { document: string, variables?: JSONObject }) => Promise<GraphqlResult<GraphqlData>>
+
+    static subscription: <GraphqlData extends JSONObject = JSONObject>({ document, variables }: { document: string, variables?: JSONObject }) => Observable<GraphqlResult<GraphqlData>>;
+}
+
+type Observable<T = object> = {
+    subscribe: (observer: Observer<T>) => {
+        unsubscribe: () => void
+    }
+}
+
+type Observer<T = object> = {
+    next?: (data: T) => void,
+    error?: (err: Error) => void,
+    done?: () => void
+}
+
+export { API };
+
+// ------ EXAMPLES -----
+
+API.put<Array<number>, { data: Array<number> }>({
```

Comment:
I think generics is better than what we're doing now. Is there any way we can do better? Opportunities to infer result types from inputs? Even if this is out of scope for this version, what about leaving surface open to allow for future inference/introspection?

Reply:
That would be cool is a worth discussion during next revision of API review post RFC